### PR TITLE
Topic/rma zcopy

### PR DESCRIPTION
**Pull request by boehms**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2343/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1677/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2418/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1762/ for details (Mellanox internal link).

File: src/ucp/core/ucp_request.h [hunk 1]
```diff
@@ -72,6 +72,7 @@ struct ucp_request {
             ucp_datatype_t        datatype; /* Send type */
             size_t                length;   /* Total length, in bytes */
             ucp_send_callback_t   cb;       /* Completion callback */
+            uct_mem_h             mem;      /* handle for zcopy */
```
Comment (yosefe) [inline src/ucp/core/ucp_request.h]: we already have it in ucp_frag_state_t

File: src/ucp/rma/basic_rma.c [hunk 2]
```diff
@@ -25,18 +25,81 @@
         return UCS_ERR_INVALID_PARAM; \
     }
 
+#define _USE_ZCOPY 1
```
Comment (yosefe) [inline src/ucp/rma/basic_rma.c]: need to make this runtime option by using zero copy threshold vairable in ucp_context

File: src/ucp/rma/basic_rma.c [hunk 3]
```diff
@@ -25,18 +25,81 @@
...
     ucs_status_t status;
     uct_rkey_t uct_rkey;
-    size_t frag_length;
     ssize_t packed_len;
     ucp_lane_index_t lane;
 
     UCP_RMA_CHECK_PARAMS(buffer, length);
 
+#if _USE_ZCOPY
+    ucp_ep_config_t *config;
```
Comment (yosefe) [inline src/ucp/rma/basic_rma.c]: already have rma_config

File: src/ucp/rma/basic_rma.c [hunk 4]
```diff
@@ -25,18 +25,81 @@
...
+        pack_ctx.src    = buffer;
+        pack_ctx.length = length;
+        packed_len = uct_ep_put_bcopy(ep->uct_eps[lane],
+                                      ucp_memcpy_pack, &pack_ctx,
+                                      remote_addr, uct_rkey);
+
+        status = (packed_len > 0) ? UCS_OK : (ucs_status_t)packed_len;
+        ucp_worker_progress(ep->worker);
+    } else {
+        uct_completion_t uct_comp;
```
Comment (yosefe) [inline src/ucp/rma/basic_rma.c]: pls declare variables in the beginning of the function

File: src/ucp/rma/basic_rma.c [hunk 5]
```diff
@@ -25,18 +25,81 @@
...
+        uct_mem_h mem = NULL;
+        uct_md_h uct_md = ucp_ep_md(ep, lane);
+
+        uct_comp.count = 2;
+
+        status = uct_md_mem_reg(uct_md, (void*)buffer, length, &mem);
+        if (status != UCS_OK) {
+            ucs_error("failed to register user buffer: %s",
+                      ucs_status_string(status));
```
Comment (yosefe) [inline src/ucp/rma/basic_rma.c]: need to return error

File: src/ucp/rma/basic_rma.c [hunk 6]
```diff
@@ -25,18 +25,81 @@
...
+                                      remote_addr, uct_rkey);
+
+        status = (packed_len > 0) ? UCS_OK : (ucs_status_t)packed_len;
+        ucp_worker_progress(ep->worker);
+    } else {
+        uct_completion_t uct_comp;
+        uct_mem_h mem = NULL;
+        uct_md_h uct_md = ucp_ep_md(ep, lane);
+
+        uct_comp.count = 2;
```
Comment (yosefe) [inline src/ucp/rma/basic_rma.c]: need to consider case where length is larger than max_zcopy supported by that UCT

File: src/ucp/rma/basic_rma.c [hunk 7]
```diff
@@ -25,18 +25,81 @@
...
+                                      uct_rkey, &uct_comp);
+
+            ucp_worker_progress(ep->worker);
+
+            if (UCS_ERR_NO_RESOURCE == status) {
+                continue;
+            }
+
+            /* Put is completed - return success */
+            status = UCS_OK;
```
Comment (yosefe) [inline src/ucp/rma/basic_rma.c]: need to check for other failures

**Comment by yosefe** [comment]
@boehms as a general comment, we need to unite blocking+nonblocking flow to one common function (probably explicit nonblocking with request) which would be called if the length>max_short (for performance reasons). In the blocking case, we would wait for the request, and in the implicit nonblocking case, we would release it.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1849/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2502/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/10/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/203/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2783/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2172/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
replaced by #1143

**Pull request closed without merging**